lib/ansible/modules/cloud/docker/docker_stack.py

```diff
@@ -0,0 +1,304 @@
+        ret[service_name] = docker_service_inspect(module, service_name)
+    return ret
+
+
+def docker_stack_rm(module, stack_name, retries, interval):
+    docker_bin = module.get_bin_path('docker', required=True)
+    command = [docker_bin, "stack", "rm", stack_name]
+
+    rc, out, err = module.run_command(command)
+
+    while err != "Nothing found in stack: %s\n" % stack_name and retries > 0:
+        sleep(interval)
+        retries = retries - 1
+        rc, out, err = module.run_command(command)
+    return rc, out, err
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec={
+            'name': dict(required=True, type='str'),
+            'compose': dict(required=False, type='list', default=[]),
+            'prune': dict(default=False, type='bool'),
+            'with_registry_auth': dict(default=False, type='bool'),
+            'resolve_image': dict(type='str', choices=['always', 'changed', 'never']),
+            'state': dict(default='present', choices=['present', 'absent']),
+            'absent_retries': dict(type='int', default=0),
+            'absent_retries_interval': dict(type='int', default=1)
+        },
+        supports_check_mode=False
+    )
+
+    if not HAS_JSONDIFF:
+        return module.fail_json(msg="jsondiff is not installed, try `pip install jsondiff`")
+
+    if not HAS_YAML:
+        return module.fail_json(msg="yaml is not installed, try `pip install pyyaml`")
+
+    state = module.params['state']
+    compose = module.params['compose']
+    name = module.params['name']
+    absent_retries = module.params['absent_retries']
+    absent_retries_interval = module.params['absent_retries_interval']
+
+    if state == 'present':
+        if not compose:
+            module.fail_json(msg=("compose parameter must be a list "
+                                  "containing at least an element"))
+
+        try:
+            compose_files = []
+            temp_files = []
+            for i, compose_def in enumerate(compose):
+                if isinstance(compose_def, dict):
+                    compose_file_fd, compose_file = tempfile.mkstemp()
+                    with os.fdopen(compose_file_fd, 'w') as stack_file:
+                        temp_files.append(compose_file)
+                        compose_files.append(compose_file)
+                        stack_file.write(yaml_dump(compose_def))
+                elif isinstance(compose_def, string_types):
+                    compose_files.append(compose_def)
+                else:
+                    module.fail_json(msg="compose %s is not a string " +
+                                     "or a dictionary" % compose_def)
+
+            before_stack_services = docker_stack_inspect(module, name)
+
+            rc, out, err = docker_stack_deploy(module, name, compose_files)
+
+            after_stack_services = docker_stack_inspect(module, name)
+
+        finally:
+            for temp_file in temp_files:
```

Comment:
`finally` blocks are not executed if `module.fail_json` is used (`fail_json()` calls `sys.exit()`)!

To make sure that a file is deleted, use `module.add_cleanup_file(filename)` after creating it.

Reply:
Ok, I removed the whole try..finally: using `module.add_cleanup_file` it is now unnecessary